File: mobile/lib/services/auth_service.dart
import 'api_client.dart';
import '../models/user.dart';

class AuthService {
  final _api = ApiClient().dio;

  Future<Map<String, dynamic>> login(String email, String password) async {
    final response = await _api.post('/auth/login', data: {
      'email': email,
      'password': password,
    });
    return response.data;
  }

  Future<Map<String, dynamic>> loginOperator(String operatorId) async {
    final response = await _api.post('/auth/login-operator/$operatorId', data: {});
    return response.data;
  }

  Future<List<User>> listOperators() async {
    final response = await _api.get('/auth/operators');
    return (response.data as List).map((e) => User.fromJson(e)).toList();
  }

  Future<User> register({
    required String fullName,
    String? email,
    String? password,
    String? photoUrl,
    String role = 'OPERATOR',
  }) async {
    final response = await _api.post('/auth/register', data: {
      'fullName': fullName,
      if (email != null) 'email': email,
      if (password != null) 'password': password,
      if (photoUrl != null) 'photoUrl': photoUrl,
      'role': role,
    });
    return User.fromJson(response.data);
  }
}

File: backend/src/auth/auth.service.ts
import {
  Injectable,
  UnauthorizedException,
  ConflictException,
  ForbiddenException,
  NotFoundException,
} from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcrypt';
import { PrismaService } from '../prisma/prisma.service';
import { LoginDto } from './dto/login.dto';
import { RegisterDto } from './dto/register.dto';
import { Role } from '@prisma/client';

@Injectable()
export class AuthService {
  constructor(
    private prisma: PrismaService,
    private jwtService: JwtService,
  ) {}

  async login(dto: LoginDto) {
    const user = await this.prisma.user.findUnique({
      where: { email: dto.email },
    });

    if (!user || !user.isActive) {
      throw new UnauthorizedException('Credenciales inválidas');
    }

    const passwordValid = await bcrypt.compare(dto.password, user.passwordHash || '');
    if (!passwordValid) {
      throw new UnauthorizedException('Credenciales inválidas');
    }

    const token = this.jwtService.sign({
      sub: user.id,
      email: user.email,
      role: user.role,
    });

    return {
      accessToken: token,
      user: {
        id: user.id,
        email: user.email,
        fullName: user.fullName,
        photoUrl: user.photoUrl,
        role: user.role,
      },
    };
  }

  async loginOperator(operatorId: string) {
    const user = await this.prisma.user.findUnique({
      where: { id: operatorId },
    });

    if (!user || !user.isActive || user.role !== Role.OPERATOR) {
      throw new NotFoundException('Operador no encontrado');
    }

    const token = this.jwtService.sign({
      sub: user.id,
      role: user.role,
    });

    return {
      accessToken: token,
      user: {
        id: user.id,
        email: user.email,
        fullName: user.fullName,
        photoUrl: user.photoUrl,
        role: user.role,
      },
    };
  }

  async listOperators() {
    return this.prisma.user.findMany({
      where: { role: Role.OPERATOR, isActive: true },
      select: {
        id: true,
        fullName: true,
        photoUrl: true,
      },
      orderBy: { fullName: 'asc' },
    });
  }

  async register(dto: RegisterDto, currentUserRole?: Role) {
    if (currentUserRole && currentUserRole !== Role.OWNER) {
      throw new ForbiddenException('Solo el dueño puede registrar usuarios');
    }

    if (dto.email) {
      const existing = await this.prisma.user.findUnique({
        where: { email: dto.email },
      });
      if (existing) {
        throw new ConflictException('El email ya está registrado');
      }
    }

    const passwordHash = dto.password ? await bcrypt.hash(dto.password, 10) : null;

    const user = await this.prisma.user.create({
      data: {
        email: dto.email || null,
        passwordHash,
        fullName: dto.fullName,
        role: dto.role || Role.OPERATOR,
      },
    });

    return {
      id: user.id,
      email: user.email,
      fullName: user.fullName,
      role: user.role,
    };
  }
}
